frontend/src/metabase/lib/api.unit.spec.ts

```diff
@@ -0,0 +1,354 @@
+          url: "/api/modified",
+        });
+
+        api.onBeforeRequestHandlers.push({
+          key: "url-modifier",
+          handler: mockHandler,
+        });
+
+        fetchMock.get("path:/api/modified", { body: { data: "modified" } });
+
+        const testEndpoint = GET("/api/test");
+        await testEndpoint();
+
+        const calls = fetchMock.callHistory.calls();
+
+        expect(calls[0].url).toContain("/api/modified");
+      });
+
+      it("should allow handlers to modify the method", async () => {
+        const mockHandler = jest.fn().mockResolvedValue({
+          method: "POST",
+        });
+
+        api.onBeforeRequestHandlers.push({
+          key: "method-modifier",
+          handler: mockHandler,
+        });
+
+        fetchMock.post("path:/api/test", { body: { data: "success" } });
+
+        const testEndpoint = GET("/api/test");
+        await testEndpoint();
+
+        const calls = fetchMock.callHistory.calls();
+
+        expect(calls[0].options.method).toBe("POST");
+      });
+
+      it("should allow handlers to modify request options by adding headers", async () => {
+        const mockHandler = jest.fn().mockResolvedValue({
+          options: {
+            headers: {
+              "X-Custom-Header": "custom-value",
+            },
+          },
+        });
+
+        api.onBeforeRequestHandlers.push({
+          key: "options-modifier",
+          handler: mockHandler,
+        });
+
+        fetchMock.get("path:/api/test-with-headers", {
+          body: { data: "success" },
+        });
+
+        const testEndpoint = GET("/api/test-with-headers");
+        await testEndpoint();
+
+        const calls = fetchMock.callHistory.calls();
+        const headers = calls[0].options.headers as Record<string, string>;
+
+        expect(headers).toMatchObject({
+          "x-custom-header": "custom-value",
+        });
+      });
+
+      it("should allow handlers to override options from invocation", async () => {
+        const mockHandler = jest.fn().mockResolvedValue({
+          options: {
+            headers: {
+              "X-Handler-Header": "handler-value",
+            },
+          },
+        });
+
+        api.onBeforeRequestHandlers.push({
+          key: "options-merger",
+          handler: mockHandler,
+        });
+
+        fetchMock.get("path:/api/test-merge-opts", {
+          body: { data: "success" },
+        });
+
+        const testEndpoint = GET("/api/test-merge-opts");
+        await testEndpoint(
+          {},
+          { headers: { "X-Original-Header": "original" } },
+        );
+
+        const calls = fetchMock.callHistory.calls();
+        const headers = calls[0].options.headers as Record<string, string>;
+
+        expect(headers).toMatchObject({
+          "x-handler-header": "handler-value",
+        });
+        // The original header from invocationOptions is overridden
+        expect(headers["x-original-header"]).toBeUndefined();
```

Comment:
Does the handler always discard existing request headers instead of merging them with new ones?